server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeShutdownAllocationDecider.java

```diff
@@ -0,0 +1,121 @@
+
+        if (thisNodeShutdownMetadata == null) {
+            // There's no shutdown metadata for this node, return yes.
+            return allocation.decision(Decision.YES, NAME, "this node is not currently shutting down");
+        }
+
+        switch (thisNodeShutdownMetadata.getType()) {
+            case REMOVE:
+                return allocation.decision(Decision.NO, NAME, "node [%s] is preparing to be removed from the cluster", node.nodeId());
+            case RESTART:
+                return allocation.decision(
+                    Decision.YES,
+                    NAME,
+                    "node [%s] is preparing to restart, but will remain in the cluster",
+                    node.nodeId()
+                );
+            default:
+                logger.error(
+                    "found unrecognized node shutdown type [{}] while deciding allocation for [{}] shard [{}][{}] on node [{}]",
+                    thisNodeShutdownMetadata.getType(),
+                    shardRouting.primary() ? "primary" : "replica",
+                    shardRouting.getIndexName(),
+                    shardRouting.getId(),
+                    node.nodeId()
+                );
+                assert false : "node shutdown type not recognized: " + thisNodeShutdownMetadata.getType();
+                return Decision.YES;
+        }
+    }
+
+    /**
+     * Applies the same rules as {@link NodeShutdownAllocationDecider#canAllocate(ShardRouting, RoutingNode, RoutingAllocation)} to
+     * determine if shards can remain on their current node.
+     */
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return this.canAllocate(shardRouting, node, allocation);
+    }
+
+    /**
+     * Prevents indices from being auto-expanded to nodes which are in the process of shutting down, regardless of whether they're shutting
+     * down for restart or removal.
+     */
+    @Override
+    public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
+        SingleNodeShutdownMetadata thisNodeShutdownMetadata = getNodeShutdownMetadata(allocation.metadata(), node.getId());
+
+        if (thisNodeShutdownMetadata == null) {
+            return allocation.decision(Decision.YES, NAME, "node [%s] is not preparing for removal from the cluster");
+        }
+
+        switch (thisNodeShutdownMetadata.getType()) {
+            case RESTART:
+                return allocation.decision(Decision.YES, NAME, "node [%s] is preparing to restart, but will remain in the cluster",
+                    node.getId());
+            case REMOVE:
+                return allocation.decision(Decision.NO, NAME, "node [%s] is preparing for removal from the cluster", node.getId());
+            default:
+                logger.error(
```

Comment:
Same comment here as above about the log spam, maybe we should just go with only the assert here?

Reply:
Same response here, I'll drop it to `DEBUG`.